### PR TITLE
Grid Layout

### PR DESCRIPTION
**Pull request by oliverlloyd**
## What does this change?
This adds support for a modern css grid layout for the article page

## Why?
The why is important here. One of the key goals for DCR is to improve the developer experience by simplifying the code to help future developers onboard easily and keep things easy to change in future. A modern grid using [grid areas](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-area) meets this goal well

## How modern css grid layout looks
Grid areas lets you display the layout of the page in a kind of ascii art format which is very easy to read

```css
                ${until.wide} {
                    grid-template-columns:
                        150px /* Left Column (220 - 1px border) */
                        1px /* Vertical grey border */
                        1fr /* Main content */
                        300px; /* Right Column */
                    grid-template-areas:
                        'title  border  headline    right-column'
                        '.      border  standfirst  right-column'
                        'meta   border  media       right-column'
                        '.      border  body        right-column';
```

## What about source foundation's grid?
Source foundation has it's [own implementation of grid](https://github.com/guardian/source-components/tree/master/packages/grid) which I was initially keen to use but after spiking it I found it did not support the use case of content spanning multiple rows or columns, such as `right-column` or `main-media` in showcase views. This precludes it from this PR.

## What about IE11?
IE11 does not support grid areas. The IE11 version is simplistic and, like source foundation, does not support content spanning multiple rows or columns. To make our layout work in IE11, with alignment between the left column and main media (the original goal for this work), we would need to have a partial grid implementation coupled with negative margins for the left column and absolute positioning for the right column. This is exactly what we have now in frontend. For this PR, instead of taking this approach I've gracefully degraded the IE11 experience, removing the left and right columns in favour of a simple single column view.

Progressive enhancement is an [established pattern](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/CSS_Grid_and_Progressive_Enhancement) for dealing with grids and whilst it detracts from the experience for IE11 users it vastly simplifies the DCR code making it easier to maintain and work with in the future. The trajectory for IE11 use is downwards and, I suspect, these users are not a target group, Indeed, it has been suggested many of them only use IE11 because it supports certain screen reader software.

### IE11 Standard
Screenshot 2020-01-20 at 13 19 42

### IE11 Showcase
Screenshot 2020-01-20 at 13 18 47

## Known issues
The 2 column view below `leftCol` but before we drop down to a single column mobile view, has a problem where there is 10px of extra spacing at the right of the page. This is being caused by the `GridItem` component for `Border` still getting rendered and `grid-column-gap: 10px;` adding space around it. You'd think it'd be simple to just hide this component for these breakpoints but for some reason this removes it from the grid for all breakpoints 🤷‍♂. More investigation is needed but I wanted to put this PR up for feedback while I do that.

## Link to supporting Trello card
https://trello.com/c/1P7ezW4F/847-align-article-meta-and-body-image

### CODE REVIEW

File: src/web/components/ArticleHeadline.tsx [hunk 1]
```diff
@@ -238,14 +231,7 @@ const renderHeadline = ({
...
                 // Inverted headlines have a wrapper div for positioning
                 // and a black background (only for the text)
-                <div
-                    className={cx(
-                        // We only shift the inverted headline down when main media is showcase
-                        isShowcase ? shiftPosition('down') : shiftSlightly,
-                        maxWidth,
-                        displayFlex,
-                    )}
-                >
```
Comment (oliverlloyd) [inline src/web/components/ArticleHeadline.tsx]: I moved this layout logic out of the headline component and put it in the layout files where it should have been to start with

File: src/web/components/ArticleTitle.tsx [hunk 2]
```diff
@@ -6,7 +6,6 @@ import { Badge } from '@frontend/web/components/Badge';
 import { SeriesSectionLink } from './SeriesSectionLink';
 
 const sectionStyles = css`
-    height: 157px;
```
Comment (oliverlloyd) [inline src/web/components/ArticleTitle.tsx]: This magic number can die now that we align automatically

File: src/web/components/Caption.tsx [hunk 3]
```diff
@@ -84,7 +86,7 @@ export const Caption: React.FC<{
     };
 
     const shouldLimitWidth =
-        role && (role === 'showcase' || role === 'supporting');
+        !isMainMedia && role && (role === 'showcase' || role === 'supporting');
```
Comment (oliverlloyd) [inline src/web/components/Caption.tsx]: This is a fix for the caption being styled as though in body when in fact it's main media. It's not actually related to grid but got boyscouted into this PR

File: src/web/components/RightColumn.tsx [hunk 4]
```diff
@@ -10,18 +10,22 @@ const hideBelowDesktop = css`
...
+        height: 100%;
         display: block;
         flex-basis: 300px;
         flex-grow: 0;
         flex-shrink: 0;
     }
 `;
 
+const hideForIE = css`
+    @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
```
Comment (oliverlloyd) [inline src/web/components/RightColumn.tsx]: Tested with browserstack using IE11 on Windows 10

File: src/web/components/ArticleHeadline.tsx (hunk 1 above)
Comment (shtukas) [reply src/web/components/ArticleHeadline.tsx]: +1

File: src/web/components/Caption.tsx (hunk 3 above)
Comment (shtukas) [reply src/web/components/Caption.tsx]: +1

File: src/web/components/RightColumn.tsx (hunk 4 above)
Comment (shtukas) [reply src/web/components/RightColumn.tsx]: +1

Comment (SiAdcock) [reply src/web/components/RightColumn.tsx]: To be clear, if we're hiding the right column for IE, does this mean the ads that normally appear in the right column at desktop and above will not appear at all for IE11 users?

Comment (oliverlloyd) [reply src/web/components/RightColumn.tsx]: Only the top most advert will no longer appear. The other in body ads will still be rendered to the right.

If we wanted to restore that top most right column advert we could patch something to absolute position that element for IE.

**Comment by gtrufitt** [comment]
> Unless we have strong evidence from product side that IE11 needs to have exact parity, I think this is fully worth it

Even if we do, I think with this we can prove it out after rolling out.

File: src/web/components/Caption.tsx (hunk 3 above)
Comment (gtrufitt) [reply src/web/components/Caption.tsx]: I guess `role` here is extra as `(role === 'showcase' || role === 'supporting')` can never be true so could drop it?

File: src/web/components/RightColumn.tsx (hunk 4 above)
Comment (gtrufitt) [inline src/web/components/RightColumn.tsx]: Could you comment a link to an explanation of this line here for prosperity?

Comment (oliverlloyd) [reply src/web/components/RightColumn.tsx]: 👍 Added SO link (it's basically feature detection but that's about as much as I understand)

Comment (oliverlloyd) [reply src/web/components/RightColumn.tsx]: 👍 Added link to SO

**Comment by rcrphillips** [comment]
Happy with this if the screen reader functionality is working.

File: src/web/components/RightColumn.tsx (hunk 4 above)
Comment (oliverlloyd) [reply src/web/components/RightColumn.tsx]: As discusses irl, we'll send out comms for this change and based on feedback decide if we need to restore this ad slot later

**Pull request merged**